frida/src/script.rs

```diff
@@ -5,46 +5,167 @@
+    pub file_name: String,
+
+    /// Line number with the error.
+    pub line_number: usize,
+
+    /// Column number with the error.
+    pub column_number: usize,
+}
+
+/// Represents a Message Log Level Types.
+/// Used by `MessageLog._level`
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+pub enum MessageLogLevel {
+    /// Indicates an informal message.
+    Info,
+    /// Represents a debugging message.
+    Debug,
+    /// Signifies a warning message.
+    Warning,
+    /// Represents an error message.
+    Error,
+}
+
+/// Represents a Message Log Level Types.
+/// Used by `MessageLog._level`
+#[derive(Deserialize, Debug)]
+pub struct SendPayload {
+    /// Send message type
+    pub r#type: String,
+    /// Send message ID
+    pub id: usize,
+    /// Send message result.
+    pub result: String,
+    /// Send message returns.
+    pub returns: Value,
+}
+
 unsafe extern "C" fn call_on_message<I: ScriptHandler>(
     _script_ptr: *mut _FridaScript,
     message: *const i8,
     _data: &frida_sys::_GBytes,
     user_data: *mut c_void,
 ) {
-    let handler: &mut I = &mut *(user_data as *mut I);
+    let c_msg = CStr::from_ptr(message as *const c_char)
+        .to_str()
+        .unwrap_or_default();
+
+    let formated_msg: Message = serde_json::from_str(c_msg).unwrap_or_else(|err| {
+        Message::Other(serde_json::json!({
+            "error": err.to_string(),
+            "data": c_msg
+        }))
+    });
 
-    handler.on_message(
-        CStr::from_ptr(message as *const c_char)
-            .to_str()
-            .unwrap_or_default(),
-    );
+    match formated_msg {
+        Message::Send(msg) => {
+            if msg.payload.r#type == "frida:rpc" {
+                let callback_handler: *mut CallbackHandler = user_data as _;
+                on_message(callback_handler.as_mut().unwrap(), Message::Send(msg));
+            }
+        }
+        Message::Log(msg) => {
+            let handler: &mut I = &mut *(user_data as *mut I);
+            handler.on_message(&Message::Log(msg));
+        }
+        Message::Error(msg) => {
+            let handler: &mut I = &mut *(user_data as *mut I);
+            handler.on_message(&Message::Error(msg));
+        }
+        Message::Other(msg) => {
+            let handler: &mut I = &mut *(user_data as *mut I);
+            handler.on_message(&Message::Other(msg));
+        }
+    }
+}
+
+fn on_message(cb_handler: &mut CallbackHandler, message: Message) {
+    let (tx, _) = &cb_handler.channel;
+    let _ = tx.send(message);
 }
 
 /// Represents a script signal handler.
 pub trait ScriptHandler {
     /// Handler called when a message is shared from JavaScript to Rust.
-    fn on_message(&mut self, message: &str);
+    fn on_message(&mut self, message: &Message);
 }
 
 /// Reprents a Frida script.
 pub struct Script<'a> {
     script_ptr: *mut _FridaScript,
     phantom: PhantomData<&'a _FridaScript>,
+    rpc_id_counter: Cell<usize>,
```

Comment:
Why use a `Cell` instead of making the necessary functions `&mut self`?

Reply:
You are right. This is done